File: go/sublist/sublist.go
// Package sublist implements routine to determine relation between 2 lists
package sublist

// Relation between 2 lists
type Relation string

// Sublist returns relation of first list with second
func Sublist(l1, l2 []int) Relation {
	switch {
	case len(l1) == len(l2) && isSublist(l1, l2):
		return "equal"
	case len(l1) < len(l2) && isSublist(l1, l2):
		return "sublist"
	case len(l2) < len(l1) && isSublist(l2, l1):
		return "superlist"
	}
	return "unequal"
}

func isSublist(l1, l2 []int) bool {
	if len(l1) == 0 {
		return true
	}

	i := indexOf(l2, l1[0], 0)
	if i == -1 || len(l1) > len(l2)-i {
		return false
	}

	for j := i; j < len(l1); j++ {
		if l1[j] != l2[i+j] {
			i = indexOf(l2, l1[0], i+1)
			if i == -1 || len(l1) > len(l2)-i {
				return false
			}
			j = -1
		}
	}

	return true
}

func indexOf(l []int, item, startIndex int) int {
	for i := startIndex; i < len(l); i++ {
		if l[i] == item {
			return i
		}
	}
	return -1
}

File: solutions/csharp/sublist/2/Sublist.cs
using System;
using System.Collections.Generic;

public enum SublistType
{
    Equal,
    Unequal,
    Superlist,
    Sublist
}

public static class Sublist
{
    public static SublistType Classify<T>(List<T> list1, List<T> list2)
        where T : IComparable
    {
        int IndexOf(List<T> list, T item, int startIndex = 0)
        {
            for (var i = startIndex; i < list.Count; i++)
            {
                return i;
            }
            return -1;
        }

        bool IsSublist(List<T> list1, List<T> list2)
        {
            if (list1.Count == 0)
            {
                return true;
            }

            var i = IndexOf(list2, list1[0]);
            if (i == -1 || list1.Count > list2.Count - i)
            {
                return false;
            }

            for (var j = 0; j < list1.Count; j++)
            {
                if (list1[j].CompareTo(list2[i + j]) != 0)
                {
                    i = IndexOf(list2, list1[0], i + 1);
                    if (i == -1 || list1.Count > list2.Count - i)
                    {
                        return false;
                    }
                    j = -1;
                }
            }
            return true;
        }

        if (list1.Count == list2.Count && IsSublist(list1, list2))
        {
            return SublistType.Equal;
        }
        else if (list1.Count < list2.Count && IsSublist(list1, list2))
        {
            return SublistType.Sublist;
        }
        else if (list1.Count > list2.Count && IsSublist(list2, list1))
        {
            return SublistType.Superlist;
        }

        return SublistType.Unequal;
    }
}